### ISSUE TITLE
Safely pass a file-like object to dask store

### ISSUE THREAD

**Issue by davidh-ssec** [dask/dask#3255]
As requested by @jakirkham I'm creating a github issue for this. I started this conversation on gitter and was about to make a SO question but have posted it here instead.

I have a custom array-like object that wraps a file-like object and I pass this custom object to Dask's `store` function to save a dask array. You can see an earlier implementation of that class [here][1] but it shouldn't matter for this question. So the issue I'm having is how to pass a file-like object as a target and have it explicitly closed. In normal execution the file seems to be closed after dask computes everything, in my unit tests with python's `unittest` the file seems to get closed before dask computes when I call it like this:

    with MyFileObject(...) as my_file:
        return da.store(my_data, my_file, compute=False)
    ...
    result.compute()  # compute the above delayed

I then decided to try using `dask.delayed` around a function that opens the file, runs store, then closes the file. But then I realized that `delayed` will compute my dask array before passing it to my function, defeating the purpose of using `da.store` for storing chunks.

So my question is, is there any way to use `da.store` with a file-like object, explicitly close that file-like object, and get the benefits of `da.store` storing individual chunks?

  [1]: https://stackoverflow.com/questions/49090399/thread-safe-rasterio-writes-from-dask-array-store

**Comment by jakirkham** [comment]
Thanks for posting this here @davidh-ssec. Sorry for taking some time getting back to you.

So hopefully this clarifies the issue. Here's basically what happens with the code above.

1. Open the file.
2. Call `da.store` with data and delayed computation.
3. `da.store` constructs a graph including the file handle and data.
4. The file handle is closed (leaving context manager scope).
5. Our function returns (leaving function scope).
6. We call `compute`.

By the time we have called `compute`, we have a graph with references to a file handle that is now closed. So it is reasonable for the code to raise an exception.

While that explains the behavior, that certainly doesn't solve the problem that you are after. So we need to think a bit about what we can do to adjust this. We have a few options. Assuming we are thinking of the single machine scheduler or threaded scheduler the three options below are all valid. If you are thinking of multiprocessing or running on many cluster jobs, only the last option is valid (unless file handles can be pickled in which case we probably need locking).

1. Move the `compute` call to within the file open/context manager block.
2. Move the  file open/context manager outside the function (and include `compute`).
3. Wrap up the file object so this open/close occurs when reading/writing data to it.

The first option is probably not satisfactory given how the example is constructed so we can drop it. The second option probably works pretty well. In this case, the function takes a file handle instead of a filename for instance. The third option also works pretty well (I use this), but may result in more code changes that you don't want to include.

My recommendation would be to try option 2 and see how it goes.

Also if you have any suggestions on how we can make this clearer to users, please let us know.

**Comment by jakirkham** [comment]
> I then decided to try using `dask.delayed` around a function that opens the file, runs store, then closes the file. But then I realized that `delayed` will compute my dask array before passing it to my function, defeating the purpose of using `da.store` for storing chunks.

As a side note, it would be interesting to know a bit more about what you were doing with `delayed`.

**Comment by davidh-ssec** [comment]
Thanks @jakirkham I'll see what I can come up with. Option 3 is probably the way I'm going to go, but it was a last resort. I'm going to spend a little time tomorrow thinking if it is possible to build a custom dask graph somehow where the last task closes the file. It can't be that hard right?

Option 1 and 2 aren't really possible with what I have planned. The package all of this logic is going in is SatPy (https://github.com/pytroll/satpy), but currently lives only in the `develop` branch and other feature branches. The current `da.store` magic is being put in our other project trollimage (https://github.com/pytroll/trollimage). The idea with setting `compute=False` on `store` is to allow satpy (and users) to merge as many parts of multiple dask graphs as possible by passing all of the Delayed objects from multiple `da.store`s to the caller who can then do `delayed(*list_of_delayeds).compute()`. Let me know if this doesn't make sense and you want more information.

My attempt at using delayed for this file stuff was to do:

```python
def _delayed(data, my_file):
    # doesn't work unless I convert data back to dask array (from_array)
    res = da.store(data, my_file)
    my_file.close()
    return res  # doesn't mean anything

delayed = dask.delayed(_delayed)(data, my_file)
if compute:
    return delayed.compute()
else:
    return delayed
```

Obviously converting it to delayed computes the entire dataset at once which isn't really desired. Right now we are only testing with threading scheduling, but you bring up a good point I had not thought of: serializing file objects. That would probably be an issue for us in the future.

**Comment by davidh-ssec** [comment]
Another thing we could do with this design is to return the data array and the target file-like object and then the user could compute it if desired. My only worry about this is that it might limit what the file creators can do. Returning a delayed object is generic, returning source data and a file is specific.

**Comment by mrocklin** [comment]
XArray runs into this problem.  I think that they open-read-close the file in every task.  This has some performance costs but works decent enough.  There was talk a while ago of caching open files.

**Comment by davidh-ssec** [comment]
Yeah I was hoping there was something like a "task wrapper" so I could put a "file open" graph task in front of a function, do the function, then put a "file close" task after that. All while somehow letting the chunks go in between. I guess at the end of the day that is the same as open-read-close the file for every chunk, just handled internally in the dask graph.

**Comment by mrocklin** [comment]
Perhaps you want a context manager?  I think my initial reaction to this is "python already has mechanisms for this, if possible we should just reuse them without designing things into dask"

**Comment by davidh-ssec** [comment]
Yeah but it is a context manager inside the dask graph essentially. Open the file once, write all the chunks, close the file once, was the goal. I'm fine moving the open/close around the individual chunk stores, but it feels like there should be something fancier...but I don't know what that is.

**Comment by mrocklin** [comment]
Nothing fancier currently exists, but there are certainly people who would
want such a thing.  I recommend that we proceed with the inefficient thing
first until it becomes a serious bottleneck.

**Comment by davidh-ssec** [comment]
Well it has become a serious bottleneck. We're using rasterio which seems to do a lot of stuff when it opens and closes files. So in a simple test case with moving the opening/closing to `__setitem__` for a block size of 4096 on an array with shape (4, 6000, 10000) it takes 5 minutes to run through all the operations and save the data to disk. With a single file open/close and computing inside the with statement it takes 30s.

@mrocklin @jakirkham Is it possible to take a delayed object and easily add another delayed function after it?

**Comment by mrocklin** [comment]
That might have more to do with a 4k blocksize than anything else.  Dask
has about 1ms of overhead.  You're going to want larger chunks.

It might be useful to try profiling to see what is taking up all of the
time.  I recommend the single-threaded scheduler and cProfile/snakeviz.
http://dask.pydata.org/en/latest/understanding-performance.html

**Comment by davidh-ssec** [comment]
A quick test and the following works:

```
            r_file = RIOFile(filename, 'r+', driver=driver,
                         width=data.sizes['x'], height=data.sizes['y'],
                         count=data.sizes['bands'],
                         dtype=dtype,
                         nodata=fill_value,
                         crs=crs, transform=transform, **format_kw)
            r_file.colorinterp = color_interp(data)
            r_file.open()
            r_file.rfile.update_tags(**new_tags)

            res = da.store(data.data, r_file, compute=compute)

            def _delayed(store_delay, r_file):
                r_file.close()
            return dask.delayed(_delayed)(res, r_file)
```

**Comment by mrocklin** [comment]
Yes, something like that would be ok.  You might run into problems if you
tried using multiple processes or machines.

**Comment by mrocklin** [comment]
We use standard python serialization protocols through pickle
https://docs.python.org/3/library/pickle.html

In particular you might implement getstate and setstate.  However it's not
clear how well this would work in your situation.  You want to call
file.close on *all* relevant workers after your tasks are done.  This isn't
a single task as is defined above, but something more complex.

I still recommend trying the open/close within each task first and
benchmarking how expensive this is likely to be.  I think that you would
need to try larger chunk sizes to get a good grasp of the relevant costs
here.

**Comment by jakirkham** [comment]
Yeah can see how a context manager would be useful. Have periodically thought about this problem for various reasons. That said, it opens us up to a lot of opportunities for things to not get closed properly when an exception is raised.

Another thing that I've been mulling over is having a function that takes all the Futures of a Dask Array. This is probably more tenable. It would take Futures as they complete and writes them out to disk. We might be able to do this on the scheduler machine if we use some sort of async implementation. Alternatively we could dispatch it as a job.

**Comment by davidh-ssec** [comment]
So this has been fun /s

Here are some timing results from playing with a few different things. Keep in mind the find resulting array that is being saved to a geotiff via rasterio is (4, 6000, 10000) of float64s. This is using the default threaded scheduler on my macbook pro (8 workers). Everything is being loaded from and saved to local disk and I'm not doing anything expensive on my system besides this. The image is made of 4 separate arrays of data so each one is pretty much saved separately, chunks of `(1, 4096, ...)`.

1. (Original) Immediate compute/close with 4096 chunk size: ~30s
2. Per chunk open/close with 4096 chunk size: ~5m
3. Using delayed close hack above: ~45s
4. Immediate compute/close with 8192 chunk size: ~34s
5. Per chunk open/close with 8192 chunk size: 2m19s to 3m
6. Using delayed close hack above: ~54s
6. Per chunk open/close with 10000 chunk size (one chunk per RGBA channel): 2m17s
7. Immediate compute/close with 10000 chunk size (one chunk per RGBA channel): 38.4s
8. Per chunk open/close with 10000 original size but rechunk to full array shape before store: 55.2s

I was not paying attention to any swapping that could have been going on so I can't comment on that. The biggest confusion for me is that adding the delayed hack seems to add a good amount of time. I didn't think so when I first tested it but I tried it multiple times and couldn't really explain it.

I've been looking at other hacky options for this and started wondering if `__dask_postcompute__` could be used for closing the file in a custom Delayed object, but theoretically that should be the equivalent of the delayed close hack, right? It's equivalent in another way that the file closing would not be working properly on a distributed scheduler.

I think if we really want the best performance we may need to change our interface so if the user/satpy asks for `compute=False` then they get the delayed object and the target file.

**Comment by mrocklin** [comment]
Can you say a bit about how you're choosing your chunk sizes?  I tend to choose chunk sizes that are in the 10MB-100MB range when I do computation.

**Comment by davidh-ssec** [comment]
@mrocklin Good question. We started by trying to guess what would be fastest so our thinking went to powers of 2 which ignores the number of bytes in the data type of the array but let's ignore that now too ;) . Then we did a couple tests to see what performed the best. As you can see from my timing 4096 seems to do better than 8192. So these are 2D arrays of 64-bit floats so if my calculations are correct that's 4096 rows x 4096 cols x 64-bit floats = ~134MB per starting dataset. iirc choosing 2048 did not perform as well as 4096 for our simple tests.

For satpy, which deals with raster data from meteorological satellites, it is very common to take a data array, the longitude for those pixels, and the latitude for those pixels and do some calculations like resampling to a uniform grid and saving the output to disk. Another common operation involves taking additional per-pixel data like solar or lunar zenith angle and doing some kind of adjustment/correction/normalization on the imagery data. Lastly, a very common operation to do with these data is to take the corrected imagery data arrays and combine them in to RGB or RGBA images.

So depending on how dask handles things like OR'ing multiple boolean arrays or any other operation that involves multiple datasets, we could end up with a couple in memory. Are you saying that our chunk size is too big or too small?

**Comment by mrocklin** [comment]
Ah, sorry, I interpreted your 4096 chunk size as being 4096 pixels, not 4096**2 pixels.  That makes sense to me.  

Does your file format support blocked storage?  When you write some block of data are you touching only one contiguous region on disk, or are you spreading over a lot of disk?  For example if your layout is C-ordered then it might make sense to rechunk your data to that ordering in order to avoid jumping around on disk a lot.  I don't know what geotiff supports here, my guess would be that it is just using naive c-ordering, in which case you probably want to pull out your data and push it back in without any chunking at all on the last axis.

**Comment by davidh-ssec** [comment]
So there are a couple things that complicate how we do this.

1. SatPy can write to multiple output formats. This issue is specifically using rasterio to save geotiffs because it was easier to save blocks instead of required the entire image to be loaded in to memory.
2. By default and pretty commonly, geotiffs are in C order I believe with one channel being contiguous at a time (all of R, then all of G, etc).
3. Geotiffs do allow tiling where a 2D region of the image is saved contiguously. This is not the default and some readers have difficulty with it...I think. We're still getting the rasterio stuff working so we haven't tackled doing this with dask blocks yet. One reason was that I assumed the best performance would be to save the blocks as is, but geotiff tiles have to be the same size (I forget how the edge tiles are handled).
4. Geotiffs can be compressed and we use DEFLATE compression by default. I'm not sure how rasterio/gdal handles this when it comes to saving blocks one at a time. Is it compressing just that block? Is it building the entire image in memory then compressing to disk? I don't know. I do know that for most of our data cases doing tiling and compression can save you a ton of disk space compared to just not tiling with compression.

**Comment by mrocklin** [comment]
Usually compressed formats are also chunked in some way.  My experience is
that if you want to get good performance storing data in chunks then you
*need* to investigate how the underlying format is doing the chunking and
match that.

**Comment by davidh-ssec** [comment]
Yeah that makes complete sense and I agree, but I would also like it to work...last week. It can get pretty complicated when it comes to GDAL and GeoTIFF given all of the options and types of compression. Looks like GDAL has a "number of threads" option for compression too: http://www.gdal.org/frmt_gtiff.html

We will definitely explore the best way to do this, but for now I wanted the basic options to work well.

**Comment by mrocklin** [comment]
I would set GDAL's number of threads for compression to 1, and rechunk your array to have no chunking along the last axes:

    x.rechunk((x.chunks[0], (x.shape[1],), (x.shape[2],)))

**Comment by jakirkham** [comment]
Missed this part before. So responding to it now.

> The idea with setting `compute=False` on `store` is to allow satpy (and users) to merge as many parts of multiple dask graphs as possible by passing all of the Delayed objects from multiple `da.store`s to the caller who can then do `delayed(*list_of_delayeds).compute()`. Let me know if this doesn't make sense and you want more information.

Typically if one wants something like this, they call `da.store` with multiple arrays to store and multiple storage targets. This combines all the Dask graphs together and is decently smart at optimizing them too. If this is something that is possible for you to do, expect that it will simplify the amount of logic carried around in your code doing similar things.

**Comment by davidh-ssec** [comment]
@jakirkham The way our package is structured you can save data to multiple different "writers" so delaying everything to the last step is preferred to reduce the amount of repeated calculations. To give these writers the most flexibility I wanted them to be able to return a generic Delayed object (or any other dask/compute-compatible object). I've made it so these individual writers can either return a single Delayed object or a `(source, target)` pair to be passed to `da.store`. The source and targets are collected and passed to one `da.store` call and the caller is responsible for closing the targets.

This is still in a pretty early version of the software so we'll see how it works out as more writers are added and used by my people.

**Comment by jakirkham** [comment]
Have you looked at the [`return_stored`]( http://dask.pydata.org/en/latest/array-api.html#dask.array.store ) argument of `store`? Setting `compute=False, return_stored=True` would give you similar behavior except it would return a Dask Array instead. This would make it a bit easier to chain together repeated calls to `store`.

**Comment by jakirkham** [comment]
Sure. The suggestion wasn't geared towards solving the file handle problem. It was geared towards solving the storing to multiple targets problem.

**Comment by davidh-ssec** [comment]
Ok, right. I had looked at it early on when looking in to the issue, but something about it didn't work with what I wanted. I had trouble understanding how saving to a file returned an array object.

**Comment by jakirkham** [comment]
The code is relatively new and has changed a bit lately. However if you have a reasonably new version of Dask, [this code is relevant]( https://github.com/dask/dask/blob/0.17.1/dask/array/core.py#L2661-L2664 ). Suggestions on how to better explain/document this functionality would be welcome.

**Comment by jakirkham** [comment]
Also if you remember what issues you ran into, those would be interesting to discuss as well.

**Comment by davidh-ssec** [comment]
If I recall correctly, I had some issues initially with performance when me and another developer were creating a test script using `da.store`. The performance issue was completely our fault, but it lead to me reading the source code for various parts of dask including the `store` function. This lead to some late night brainstorming of "what if every writer used `store`...". So maybe we could add documentation to dask with an example of using `store` in some way similar to these (file-like/array-like object as a target). I think a lot of my questions and this issue are mostly results of me being over curious about dask's inner workings and wanting to optimize my code early.

**Comment by davidh-ssec** [comment]
I've been watching some dask presentation videos today which got me thinking about this problem again.

1. Is there a way to detect which scheduler is being used or will be used while building the dask graph? My thought being if it is threaded then I can open the file once and share that between the threads (with locking). If it is multiprocess then it could use the open-file-per-chunk method (assuming that is allowed by the file format).
2. Is there a way to force a set of chunks to all compute on the same thread/worker? The documentation [here](https://distributed.readthedocs.io/en/latest/faq.html#does-dask-handle-data-locality) says "You can tell Dask to preferentially run a task on a particular worker or set of workers." which leads me to the source code. I'm hoping I can get the short answer here.
3. Is it possible to do complex schedulers like multiprocess with multiple threads? This may not be super useful but I'm thinking of opening a file whose writing library is thread-safe but not multiprocess safe and forcing the write tasks for a single file to one worker process but use multiple worker threads.
4. Is is possible to get anything realtime/live when working with the threaded scheduler? I'm preparing a presentation for SciPy this year (thanks for reviewing Matt) and if I can't get file writing to be completely multi-process friendly then my visuals might not be as exciting. Haven't thought about too many other options either.

**Comment by davidh-ssec** [comment]
@jakirkham 2) My use case in SatPy involves combining one or more input files together in various combinations and saving those combinations to multiple files, one file per combination (RGB images), possibly multiple output formats. The final tasks in all of these graphs is writing to the file. If opening the file in multiple processes is an issue for the file format then limiting each file write task (one task, multiple chunks) to a single process then I could open the file once in that worker and not worry about closing it until the end. Similarly for non-thread-safe file formats I could have one worker thread be responsible for writing all the chunks of one file. Hopefully that made some kind of sense.

3)  Yes I suppose it is switching between schedulers. I'm not sure how to easily hide that from the user in my library, but I'll keep it in mind. If you combined point 2 above and this point it might make it easier to understand. Definitely a lot of good stuff to read.

4) I'll take a look. Thanks.

**Comment by jakirkham** [comment]
Thanks for the clarification.

Yes 2 makes sense now and can see why that would be nice. So [this might help]( http://distributed.readthedocs.io/en/latest/locality.html#specify-workers-with-compute-persist ).

There are things like `temp_default_client`, which can be nice here. It's technically private API though. ( https://github.com/dask/distributed/pull/1849 ). I guess in most real world scenarios a specific `Client` object is always used. So can create a `Client` and just use it internally and remove it when done.

**Comment by davidh-ssec** [comment]
@jakirkham I may have to open another issue up for this, but it's also a semi-expected result. When I use `Client(processes=False)` I get an exception during computation about serialization of an object. I know the object isn't serializable, but I thought this was supposed to leave everything threaded?

Also last time I tried this without the `processes` keyword just to see what would work I don't get any serialization exceptions but do get errors from the file library. I'll debug further, but let me know if I'm misunderstanding something.

**Comment by jakirkham** [comment]
What sort of serialization are we referring to? Are we discussing writing to a file or pickling/transferring data?

IIUC `processes=False` should only use threads (no processes whatsoever).

That said, Python threads are just POSIX threads (on POSIX compliant systems like Unix). They can run at the same time if the GIL ends up being released, which often happens with IO. So you may need to do some locking if you are running into problems serializing data to disk.

**Comment by davidh-ssec** [comment]
I was getting serialization errors from it trying to serialize the file objects it was going to save to. The locking for this code is handled by `da.store` iirc. It seems like things were being serialized when `processes=False`.

**Comment by jakirkham** [comment]
Failed to xref this, but had raised issue ( https://github.com/dask/distributed/issues/2056 ) on the serialization issue with Distributed. Basically it seems we always pickle even in the same process.

**Comment by mrocklin** [comment]
If you wrap things in dask.delayed you should be ok

**Comment by mrocklin** [comment]
Any update on this @djhoese ?

**Comment by djhoese** [comment]
@mrocklin No real update. I will test what happens with rasterio (saving to geotiff) and using the local client. I don't expect much but maybe things have changed a bit. Based on the series of issues that have been linked it looks like there are things that might make this easier in the future, but there still isn't an "approved" pattern for passing file-like objects to dask. Something I just thought of would be something like (forgive me, 5 minutes of thought):

```python
from dask import DelayedFile
# (open_callback, open_args, close_callback, close_args)
dfile = DelayedFile(open, 'my_file.dat', 'close', tuple())
da.store(my_dask_arr, dfile)
```

Then *magically* dask opens the file when the chunks are ready to be written to, closed when done being written to. Additionally, if the scheduler is threaded then it opens the file once and uses it for all workers.

**Comment by shoyer** [comment]
We recently wrote a CachingFileManager object to handle this sort of logic in xarray: 
https://github.com/pydata/xarray/blob/master/xarray/backends/file_manager.py

We could potentially upstream this to dask if desired.

**Comment by djhoese** [comment]
Thanks @shoyer I'll have to give that a try.

**Comment by djhoese** [comment]
@shoyer How does xarray currently use the `CachingFileManager`? Do you currently have any dask components calling the `acquire` method? I'm going to try subclassing `CachingFileManager` and creating a `__setitem__` so that I can pass it to `da.store`. I'm traveling a lot this weekend but I'll keep this issue updated.

**Comment by shoyer** [comment]
@djhoese take a look at the code for our rasterio backend, it should be pretty self explanatory: 
https://github.com/pydata/xarray/blob/master/xarray/backends/rasterio_.py

Please don’t subclass CachingFileManager, but it should work fine to save it as an attribute of another object that gets pickled (e.g., by dask-distributed). That’s what we’re doing in xarray.

**Comment by scottyhq** [comment]
@djhoese, @mrocklin, and @shoyer  I'm very interested in getting rasterio's WarpedVRT to work with `xr.open_rasterio()`. The discussion here seems very relevant, but happy to open a new issue (not sure if xarray, dask or rasterio is the best place!

I Recently took a stab at modifying `rasterio_.py` but am running into errors such as:

```
distributed.protocol.core - CRITICAL - Failed to Serialize
Traceback (most recent call last):
  File "/Users/scott/miniconda3/envs/test_env/lib/python3.6/site-packages/distributed/protocol/core.py", line 53, in dumps
    for key, value in data.items()
  File "/Users/scott/miniconda3/envs/test_env/lib/python3.6/site-packages/distributed/protocol/core.py", line 54, in <dictcomp>
    if type(value) is Serialize}
  File "/Users/scott/miniconda3/envs/test_env/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 157, in serialize
    raise TypeError(msg, str(x)[:10000])
TypeError: ('Could not serialize object of type ImplicitToExplicitIndexingAdapter.', 'ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyOuterIndexedArray(array=<xarray.backends.rasterio_.RasterioArrayWrapper object at 0x10ac6b208>, key=BasicIndexer((slice(None, None, None), slice(None, None, None), slice(None, None, None))))))')
distributed.comm.utils - ERROR - ('Could not serialize object of type ImplicitToExplicitIndexingAdapter.', 'ImplicitToExplicitIndexingAdapter(array=CopyOnWriteArray(array=LazilyOuterIndexedArray(array=<xarray.backends.rasterio_.RasterioArrayWrapper object at 0x10ac6b208>, key=BasicIndexer((slice(None, None, None), slice(None, None, None), slice(None, None, None))))))')
```

In brief, currently we can currently do this:
```python
image_url = 'https://storage.googleapis.com/gcp-public-data-landsat/LC08/01/047/027/LC08_L1TP_047027_20130421_20170310_01_T1/LC08_L1TP_047027_20130421_20170310_01_T1_B4.TIF'

xchunk = 2048
ychunk = 2048
with rasterio.open(image_url) as src:
        da = xr.open_rasterio(src.name, 
                         chunks={'band': 1, 'x': xchunk, 'y': ychunk},
                         )
        print(da)
        da.isel(band=0, x=slice(0,2000), y=slice(0,3000)).plot.imshow(vmin=-150, vmax=150, center=False)
```

But NOT this:
```python
with rasterio.open(image_url) as src:
    with WarpedVRT(src, crs='+init=epsg:3857 +no_defs') as vrt:
        da = xr.open_rasterio(vrt.name, 
                         chunks={'band': 1, 'x': xchunk, 'y': ychunk},
                         )
        print(da)
        da.isel(band=0, x=slice(0,2000), y=slice(0,3000)).plot.imshow(vmin=-150, vmax=150, center=False)
```

In-memory “virtual” reprojection is a key feature of rasterio (see https://gist.github.com/sgillies/7e5cd548110a5b4d45ac1a1d93cb17a3), and it would be fantastic if this worked w/ xarray and dask distributed. This is a very common need, especially between WGS84 lat/lon, UTM, Google Mercator.

**Comment by shoyer** [comment]
@scottyhq let's discuss this over in the xarray tracker. This would probably be easiest to diagnose if you can share a pull request with your attempted changes.

**Comment by mrocklin** [comment]
I'm curious to follow the discussion.  Links from here are welcome

**Comment by LTMeyer** [comment]
I believe I am facing a similar issue as @djhoese, that's why I'm posting it here. I'm happy to post it anywhere it would be more relevant.

## Context ##

I am trying to compute NDVI values from a large multi-spectral geotiff. The different bands are extracted from a local file and the resulting NDVI band written in another file. For the purpose I was interested in using `dask` and `rasterio`. Inspired by [dask example](https://examples.dask.org/applications/satellite-imagery-geotiff.html) and @mrocklin's [tutorial](http://matthewrocklin.com/blog/work/2017/01/17/dask-images) I am attempting to load in parallel blocks from the geotiff input and generate NDVI values.

## Code Sample ## 

```python
def compute_ndvi(block):
    """Compute the NDVI from a block that contains all bands. 
    Block is expected to be a 3D array with the first dimension representing the band.

    """
    if block.shape[0] == 0:
        return block
    red = block[0, :, :]
    nir = block[4, :, :]
    ndvi = (nir - red) / (nir + red)
    return ndvi

with rasterio.open(path_to_my_local_file) as src:
        profile = src.profile
        block_shapes = src.block_shapes[0]  # Give the block shapes for each band of the raster
        ds = xr.open_rasterio(src, chunks={"x":block_shapes[0], "y":block_shapes[1]})
        ndvi = da.map_blocks(compute_ndvi, ds)
        profile.update(count=1, compress="lzw")
        with rasterio.open("ndvi.tif", 'w', **profile) as dst:
            da.store(ndvi, dst, compute=True)
```

The code above raises the following error:
> TypeError: self._hds cannot be converted to a Python object for pickling

## Related Issues ##

My rough understanding is that there is an issue in parallel writing into a geotiff through `rasterio`. Issue for which I believe @djhoese has done more exhaustive research, hence has a better understanding.

This [SO question](https://stackoverflow.com/questions/49090399/thread-safe-rasterio-writes-from-dask-array-store) makes me think such behaviour was working in the past, which is not the case anymore. Besides, similar issue is mentioned in rasterio repository (https://github.com/mapbox/rasterio/issues/1731). I haven't, however, fully understood the suggested solution and I'm afraid it undermine the performance gain brought by dask's parallelisation. 

Could someone with a better understanding point me to a working solution. In case such solution would not be existing yet, could someone outline the underlying problems and how to start tackling them?

**Comment by djhoese** [comment]
What is the full traceback you are getting? Looking back at my SO question (yes that's me too) you will need to at least specify `lock=True` in your `da.store` call.

Although I think it is supposed to work now, you should be careful passing xarray objects to dask's functions. Instead you should pass the underlying dask array `ds.data`.

**Comment by LTMeyer** [comment]
Thank you for your quick answer.

> What is the full traceback you are getting? Looking back at my SO question (yes that's me too) you will need to at least specify `lock=True` in your `da.store` call.

I've added the full error log to the message above.
Regarding the `lock=True` option, I've tried with and without. It didn't bring better success. Actually I believe the option is set to `True` by default (https://docs.dask.org/en/latest/array-api.html#dask.array.store).
> Although I think it is supposed to work now, you should be careful passing xarray objects to dask's functions. Instead you should pass the underlying dask array `ds.data`.

Where should I pass the underlying data?

> Edit: And do you have dask configured specially at all? Is it using the default scheduler (threaded) or are you trying to use a distributed/multiprocess scheduler?

Unfortunately I am a total beginner with `dask`. Hence, I haven't set any configuration yet. Would multiprocessing solve the issue?

**Comment by djhoese** [comment]
> This probably needs to be done either in some separate function, or in a class with a __getitem__ accessor that doe this, and then that class could then be passed into store.
> 
> Xarray does this. That seems like the easiest approach short term.

I agree with @mrocklin. The class/solution I have in my SO question is the same solution I'm using right now in the `trollimage` package, but I've always meant to look deeper into how xarray is doing it under the hood. My class in that SO question has a custom `__setitem__` to support this windowed writing.

> Where should I pass the underlying data?

You'll need to pay attention to what type of objects your variables are. For example,

```
        ndvi = da.map_blocks(compute_ndvi, ds)
        profile.update(count=1, compress="lzw")
        with rasterio.open("ndvi.tif", 'w', **profile) as dst:
            da.store(ndvi, dst, compute=True)```

Your `ds` at the start is an xarray DataArray with a dask array underneath. So `ds` is DataArray, `ds.data` is dask.array.Array. This may have been fixed recently, but I would consider it best practice to pass only dask arrays to dask functions like `da.map_blocks` and `da.store`. So in your `map_blocks` call you should pass `ds.data`, not `ds`.

In your `da.store` call you are passing a rasterio object. I don't think rasterio objects support sliced writing via `__setitem__` which is what `da.store` is going to try to use. You'd need to write your own custom class to wrap that like I did in the SO question (trollimage package) or look at how xarray is doing it underneath (spoiler: it is probably the same `__setitem__` approach).

**Comment by djhoese** [comment]
I should have also said: Dask uses the threaded scheduler by default. The multiprocessing and distributed based schedulers will make things more difficult for you as far as rasterio is concerned. In those cases you'd have to do what Matt suggests and try to close the created file, reopen it on every worker, then write the chunk of data you are writing, but I'm not sure rasterio supports this type of operation.

**Comment by LTMeyer** [comment]
Thank you both @mrocklin and @djhoese. I understand the SO was already giving the answer. I will try that and tell you if it worked.

Additionally, I tried to check with scheduler I was using:
```python
dask.config.get("scheduler")
>>> 'dask.distributed'
```
Not sure it what we expected.
On the long term however I would like to possibly involve several remote machines, but it is not clear yet.

**Comment by djhoese** [comment]
Are you creating a `Client` object before running the code you pasted originally? Or are you running on a pre-built environment like something running on a binderhub? I'd expect the distributed scheduler you've found in these cases, but it is not typically the default. If you are creating the `Client` and aren't sure you need it, then I would remove it.

**Comment by LTMeyer** [comment]
My bad, I am indeed a client: through the dask jupyterlab extension to understand what is happening. I followed the instruction in this video (https://youtu.be/EX_voquHdk0?t=176).
However, I don't believe this client is configured itself. It must have the default values set by the extension.

**Comment by shoyer** [comment]
> The class/solution I have in my SO question is the same solution I'm using right now in the `trollimage` package, but I've always meant to look deeper into how xarray is doing it under the hood.

The main extra trick we use is a (global) LRU cache for open files within each process, which lets us significantly reduce overhead. This machinery is relatively isolated, and we might be able to factor it out in the upcoming xarray backends refactor. At the very least it would be nice to expose it as a documented external API.

**Comment by mrocklin** [comment]
> The main extra trick we use is a (global) LRU cache for open files within each process, which lets us significantly reduce overhead. This machinery is relatively isolated, and we might be able to factor it out in the upcoming xarray backends refactor. At the very least it would be nice to expose it as a documented external API.

This comes up often enough for dask array (mostly with HDF5) that we should probably have our own solution in the future.  The answer of "you should probably go use Xarray, and then pull out the `.data` attribute if you really want a Dask Array" is probably not entirely ideal (although it's taken us fairly far).

**Comment by LTMeyer** [comment]
> I should have also said: Dask uses the threaded scheduler by default. The multiprocessing and distributed based schedulers will make things more difficult for you as far as rasterio is concerned. In those cases you'd have to do what Matt suggests and try to close the created file, reopen it on every worker, then write the chunk of data you are writing, but I'm not sure rasterio supports this type of operation.

The solution for the threaded scheduler works indeed. Nonetheless, I would be keen to make it work with multiprocessing too. According to [rasterio documentation](https://rasterio.readthedocs.io/en/latest/topics/concurrency.html) it is possible too.

**Comment by djhoese** [comment]
I'm very confused by that rasterio example. I've never used a ThreadPoolExecutor, but it doesn't seem like any of the multi-threaded/parallel code in that example has to do with rasterio. The only thing spanning thread boundaries are numpy arrays that are read from a geotiff in the main thread and written to a geotiff in the main thread. The only thing that looks possibly multi-threaded for rasterio is:

```
                    for window, result in zip(
                        windows, executor.map(compute, data_gen)
                    ):
```

But that would mean that `.map` is passing generator context to each thread which seems dangerous rather than iterating over the generator and giving the resulting object (a numpy array) to the `compute` function in a separate thread.
